File: internal/vens/session.go
package vens

import (
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"io"
	"log/slog"
	"net"
	"time"
)

// ControlSession manages TCP control channel connections (port 53219).
// Each operation opens a new TCP connection, following the scanner protocol.
type ControlSession struct {
	host string
	port uint16
}

// NewControlSession creates a ControlSession for the given scanner address.
func NewControlSession(host string, port uint16) *ControlSession {
	return &ControlSession{host: host, port: port}
}

// connect opens a TCP connection and reads the welcome packet.
func (s *ControlSession) connect() (net.Conn, error) {
	addr := net.JoinHostPort(s.host, fmt.Sprintf("%d", s.port))
	slog.Debug("control channel connecting", "addr", addr)
	conn, err := net.DialTimeout("tcp", addr, 5*time.Second)
	if err != nil {
		return nil, fmt.Errorf("control connect: %w", err)
	}
	conn.SetDeadline(time.Now().Add(10 * time.Second))

	welcome := make([]byte, WelcomeSize)
	if _, err := io.ReadFull(conn, welcome); err != nil {
		conn.Close()
		return nil, fmt.Errorf("control welcome: %w", err)
	}
	if err := ValidateWelcome(welcome); err != nil {
		conn.Close()
		return nil, err
	}
	slog.Debug("received welcome from control channel", "addr", addr, "bytes", len(welcome), "hex", hex.EncodeToString(welcome))
	return conn, nil
}

// sendRecv opens a connection, sends data, reads a length-prefixed response, and closes.
func (s *ControlSession) sendRecv(data []byte) ([]byte, error) {
	conn, err := s.connect()
	if err != nil {
		return nil, err
	}
	defer conn.Close()

	slog.Debug("control send", "bytes", len(data), "hex", hex.EncodeToString(data))
	if _, err := conn.Write(data); err != nil {
		return nil, fmt.Errorf("control send: %w", err)
	}

	// Read response length (first 4 bytes)
	lenBuf := make([]byte, 4)
	if _, err := io.ReadFull(conn, lenBuf); err != nil {
		return nil, fmt.Errorf("control recv length: %w", err)
	}
	respLen := binary.BigEndian.Uint32(lenBuf)

	resp := make([]byte, respLen)
	copy(resp[:4], lenBuf)
	if _, err := io.ReadFull(conn, resp[4:]); err != nil {
		return nil, fmt.Errorf("control recv body: %w", err)
	}
	slog.Debug("control recv", "bytes", len(resp), "hex", hex.EncodeToString(resp))
	return resp, nil
}

// Register registers this client with the scanner.
func (s *ControlSession) Register(token [8]byte) error {
	slog.Debug("registering with scanner...")
	conn, err := s.connect()
	if err != nil {
		return err
	}
	defer conn.Close()

	req := MarshalReleaseRequest(token, 1)
	slog.Debug("register request", "bytes", len(req), "hex", hex.EncodeToString(req))
	if _, err := conn.Write(req); err != nil {
		return fmt.Errorf("register send: %w", err)
	}

	// Register response is a fixed 16-byte ack
	ack := make([]byte, 16)
	if _, err := io.ReadFull(conn, ack); err != nil {
		return fmt.Errorf("register recv: %w", err)
	}
	slog.Debug("registration response", "bytes", len(ack), "hex", hex.EncodeToString(ack))
	return nil
}

// Configure sends client configuration (identity, notify port, etc.) to the scanner.
// Returns true if the scanner accepted the pairing.
func (s *ControlSession) Configure(token [8]byte, clientIP string, notifyPort uint16, identity string) (bool, error) {
	req := MarshalReserveRequest(token, clientIP, notifyPort, identity, time.Now())
	slog.Debug("configuring session", "ip", clientIP, "port", notifyPort, "identity_len", len(identity))

	resp, err := s.sendRecv(req)
	if err != nil {
		return false, err
	}

	slog.Debug("configure response", "bytes", len(resp))
	status, err := ParseReserveResponse(resp)
	if err != nil {
		return false, err
	}
	if status == 0 {
		slog.Info("pairing accepted")
		return true, nil
	}
	slog.Info("pairing rejected", "status", status)
	return false, nil
}

// CheckStatus queries the scanner's connection status.
func (s *ControlSession) CheckStatus(token [8]byte) (uint32, error) {
	slog.Debug("checking scanner status...")
	req := MarshalGetWifiStatusRequest(token)
	resp, err := s.sendRecv(req)
	if err != nil {
		return 0, err
	}
	state, err := ParseGetWifiStatusResponse(resp)
	if err != nil {
		return 0, err
	}
	slog.Debug("status check", "state", state, "response_bytes", len(resp))
	return state, nil
}

// Deregister removes this client from the scanner.
func (s *ControlSession) Deregister(token [8]byte) error {
	slog.Debug("deregistering...")
	conn, err := s.connect()
	if err != nil {
		return err
	}
	defer conn.Close()

	req := MarshalReleaseRequest(token, 0)
	slog.Debug("deregister request", "bytes", len(req), "hex", hex.EncodeToString(req))
	if _, err := conn.Write(req); err != nil {
		return fmt.Errorf("deregister send: %w", err)
	}

	ack := make([]byte, 16)
	if _, err := io.ReadFull(conn, ack); err != nil {
		return fmt.Errorf("deregister recv: %w", err)
	}
	slog.Debug("deregistration response", "bytes", len(ack), "hex", hex.EncodeToString(ack))
	return nil
}

File: scansnap/session.py
"""TCP control channel (port 53219) — session management."""

from __future__ import annotations

import asyncio
import logging
import struct
from datetime import datetime

from scansnap.packets import (
    CLIENT_NOTIFY_PORT,
    ReserveRequest,
    ReleaseRequest,
    GetWifiStatusRequest,
    GetWifiStatusResponse,
    WelcomePacket,
)

log = logging.getLogger(__name__)


async def _read_exact(reader: asyncio.StreamReader, n: int) -> bytes:
    """Read exactly n bytes from stream."""
    buf = bytearray()
    while len(buf) < n:
        chunk = await reader.read(n - len(buf))
        if not chunk:
            raise ConnectionError("Connection closed while reading")
        buf.extend(chunk)
    return bytes(buf)


class ControlSession:
    """Manages a TCP control channel connection to the scanner."""

    def __init__(self, host: str, port: int) -> None:
        self.host = host
        self.port = port
        self._reader: asyncio.StreamReader | None = None
        self._writer: asyncio.StreamWriter | None = None

    async def _connect(self) -> tuple[asyncio.StreamReader, asyncio.StreamWriter]:
        reader, writer = await asyncio.open_connection(self.host, self.port)
        # Read the welcome packet
        welcome_data = await _read_exact(reader, WelcomePacket.size())
        WelcomePacket.unpack(welcome_data)
        log.debug("Received welcome from %s:%d", self.host, self.port)
        return reader, writer

    async def _send_recv(self, data: bytes) -> bytes:
        """Open a new connection, send data, read response, close."""
        reader, writer = await self._connect()
        try:
            writer.write(data)
            await writer.drain()
            # Read response length first (4 bytes)
            len_data = await _read_exact(reader, 4)
            resp_len = int.from_bytes(len_data, "big")
            rest = await _read_exact(reader, resp_len - 4)
            return len_data + rest
        finally:
            writer.close()
            await writer.wait_closed()

    async def register(self, token: bytes) -> bytes:
        """Register with the scanner. Returns the raw response (16-byte ack)."""
        req = ReleaseRequest(token=token, action=1)
        log.info("Registering with scanner...")
        reader, writer = await self._connect()
        try:
            writer.write(req.pack())
            await writer.drain()
            # Register response is a 16-byte ack, not length-prefixed
            resp = await _read_exact(reader, 16)
            log.info("Registration response: %d bytes, hex=%s", len(resp), resp.hex())
            return resp
        finally:
            writer.close()
            await writer.wait_closed()

    async def configure(
        self,
        token: bytes,
        client_ip: str,
        notify_port: int = CLIENT_NOTIFY_PORT,
        identity: str = "",
    ) -> bytes:
        """Send client configuration to the scanner."""
        req = ReserveRequest(
            token=token,
            client_ip=client_ip,
            notify_port=notify_port,
            identity=identity,
            timestamp=datetime.now(),
        )
        log.info("Configuring session (ip=%s, port=%d)...", client_ip, notify_port)
        resp = await self._send_recv(req.pack())
        log.info("Configure response: %d bytes", len(resp))
        return resp

    async def try_configure(
        self,
        token: bytes,
        client_ip: str,
        notify_port: int = CLIENT_NOTIFY_PORT,
        identity: str = "",
    ) -> bool:
        """Send ReserveRequest and return True if accepted, False if rejected."""
        resp = await self.configure(token, client_ip, notify_port, identity)
        status = struct.unpack_from("!i", resp, 8)[0]
        if status == 0:
            log.info("Pairing accepted")
            return True
        log.info("Pairing rejected (status=%d)", status)
        return False

    async def check_status(self, token: bytes) -> GetWifiStatusResponse:
        """Check connection status."""
        req = GetWifiStatusRequest(token=token)
        resp = await self._send_recv(req.pack())
        status = GetWifiStatusResponse.unpack(resp)
        log.debug("Status: state=%d", status.state)
        return status

    async def deregister(self, token: bytes) -> bytes:
        """Deregister from the scanner."""
        req = ReleaseRequest(token=token, action=0)
        log.info("Deregistering...")
        reader, writer = await self._connect()
        try:
            writer.write(req.pack())
            await writer.drain()
            resp = await _read_exact(reader, 16)
            log.info("Deregistration response: %d bytes", len(resp))
            return resp
        finally:
            writer.close()
            await writer.wait_closed()
